File: filesync/src/main/java/de/mel/filesync/data/conflict/Conflict.java
package de.mel.filesync.data.conflict;

import de.mel.Lok;
import de.mel.auth.tools.Eva;
import de.mel.filesync.sql.Stage;
import de.mel.filesync.sql.dao.StageDao;

import java.util.*;

/**
 * Created by xor on 5/30/17.
 */
public class Conflict {

    private Stage lStage;
    private Stage rStage;
    private Long lStageId, rStageId;
    private StageDao stageDao;
    private Boolean isRight;
    private String key;
    private Set<Conflict> dependents = new HashSet<>();
    private Conflict dependsOn;

    public Conflict(StageDao stageDao, Stage lStage, Stage rStage) {
        this.lStageId = lStage != null ? lStage.getId() : null;
        this.rStageId = rStage != null ? rStage.getId() : null;
        this.lStage = lStage;
        this.rStage = rStage;
        this.stageDao = stageDao;
        key = createKey(lStage, rStage);
        if (lStageId != null && rStageId != null && lStageId == 18 && rStageId == 40)
            Lok.debug("Conflict.Conflict");
        Eva.eva((eva, count) -> {
            if (lStageId != null && rStageId != null && lStageId == 18 && rStageId == 40)
                eva.error();
        });
    }


    public Conflict() {

    }

    public static String createKey(Stage lStage, Stage rStage) {
        return (lStage != null ? lStage.getId() : "n") + "/" + (rStage != null ? rStage.getId() : "n");
    }

    public Conflict chooseRight() {
        isRight = true;
        for (Conflict sub : dependents) {
            sub.chooseRight();
        }
        return this;
    }

    public Conflict chooseLeft() {
        isRight = false;
        for (Conflict sub : dependents) {
            sub.chooseLeft();
        }
        return this;
    }

    public boolean isRight() {
        return isRight != null && isRight;
    }

    /**
     * get you the stage that was chosen. that might be a directory higher in the hierarchy.
     * it isn't necessarily the parent stage.
     * @return
     */
    public Stage getChoice() {
        if (isRight == null)
            return null;
        if (isRight) {
            if (rStage != null)
                return rStage;
            else
                return dependsOn.getChoice();
        } else {
            if (lStage != null)
                return lStage;
            else
                return dependsOn.getChoice();
        }
    }

    public boolean hasDecision() {
        return isRight != null;
    }

    public String getKey() {
        return key;
    }

    public Stage getLeft() {
        if (lStageId == null)
            return null;
        return lStage;
    }

    public Stage getRight() {
        if (rStageId == null)
            return null;
        return rStage;
    }

    @Override
    public String toString() {
        return "{class:\"" + getClass().getSimpleName() + "\",key:\"" + key + "\",isRight:\"" + isRight + "\"}";
    }


    public void chooseNothing() {
        isRight = null;
    }

    public boolean isLeft() {
        return isRight != null && !isRight;
    }

    public Conflict dependOn(Conflict dependsOn) {
        this.dependsOn = dependsOn;
        if (dependsOn != null)
            dependsOn.dependents.add(this);
        return this;
    }

    public Conflict getDependsOn() {
        return dependsOn;
    }

    public Set<Conflict> getDependents() {
        return dependents;
    }

    public boolean hasLeft() {
        return lStageId != null;
    }

    public boolean hasRight() {
        return rStageId != null;
    }

    /**
     * sorts, indents and adds empty rows
     *
     * @param conflicts
     * @return
     */
    public static List<Conflict> prepareConflicts(Collection<Conflict> conflicts) {
        List<Conflict> result = new ArrayList<>();
        List<Conflict> rootConflicts = new ArrayList<>();
        Set<Conflict> independentConflicts = new HashSet<>();
        for (Conflict conflict : conflicts) {
            if (conflict.getDependsOn() == null)
                rootConflicts.add(conflict);
            else if (conflict.getDependents().size() == 0)
                independentConflicts.add(conflict);
        }
        for (Conflict root : rootConflicts) {
            result.add(root);
            traversalAdding2(result, root.getDependents());
            if (root.getDependents().size() > 0)
                result.add(new EmptyRowConflict());
        }
        return result;
    }

    private static void traversalAdding2(List<Conflict> result, Set<Conflict> stuffToTraverse) {
        for (Conflict conflict : stuffToTraverse) {
            result.add(conflict);
            if (conflict.getDependents().size() > 0) {
                traversalAdding2(result, conflict.getDependents());
            }
        }
    }

//    /**
//     * sorts, indents and adds empty rows
//     *
//     * @param conflicts
//     * @return
//     */
//    public static List<Conflict> prepareConflicts(Collection<Conflict> conflicts) {
//        List<Conflict> result = new ArrayList<>();
//        List<Conflict> rootConflicts = getRootConflicts(conflicts);
//        for (Conflict root : rootConflicts) {
//            result.add(root);
//            traversalAdding(result, root.getDependents());
//            if (root.getDependents().size() > 0)
//                result.add(new EmptyRowConflict());
//        }
//        return result;
//    }
//
//    private static void traversalAdding(List<Conflict> result, Set<Conflict> stuffToTraverse) {
//        for (Conflict conflict : stuffToTraverse) {
//            result.add(conflict);
//            if (conflict.getDependents().size() > 0) {
//                traversalAdding(result, conflict.getDependents());
//            }
//        }
//    }

    public Conflict getDependentByName(String name) {
        for (Conflict dep : dependents) {
            if (dep.getLeft() != null && dep.getLeft().getName().equals(name))
                return dep;
            if (dep.getRight() != null && dep.getRight().getName().equals(name))
                return dep;
        }
        return null;
    }

    public static List<Conflict> getRootConflicts(Collection<Conflict> conflicts) {
        List<Conflict> rootConflicts = new ArrayList<>();
        for (Conflict conflict : conflicts) {
            if (conflict.getDependsOn() == null)
                rootConflicts.add(conflict);
        }
        return rootConflicts;
    }
}

File: filesync/src/main/java/de/mel/filesync/data/conflict/Conflict.kt
package de.mel.filesync.data.conflict

import de.mel.Lok
import de.mel.filesync.sql.Stage
import de.mel.filesync.sql.dao.ConflictDao
import java.util.Observable

/**
 * Basic object that finds pairs of local a remote conflicts. Conflicts are based on [Stage]s of a local and a remote StageSet.
 * So it may happen that only one side has a [Stage] because it does not exist in the other.
 */
open class Conflict(val conflictDao: ConflictDao, val localStage: Stage?, val remoteStage: Stage?) {
    private var localId: Long?
    private var remoteId: Long?

    var decision: Stage? = null
        get

    var rejection: Stage? = null
        get
    var key: String
        get
    var parent: Conflict? = null
        get
    var children = mutableListOf<Conflict>()
    var debugChildren  = mutableSetOf<Conflict>()


    var chosenLocal: Boolean = false
        get() = this.hasChoice && this.decision === localStage
    var chosenRemote: Boolean = false
        get() = this.hasChoice && this.decision === remoteStage
    val hasChoice: Boolean
        get() = this.decision != null || this.rejection != null

    init {
        key = createKey(localStage, remoteStage)
        localId = localStage?.id
        remoteId = remoteStage?.id
    }

    fun dependOn(parent: Conflict?) {
        this.parent = parent
        parent?.children?.add(this)
    }

    fun addChild(child: Conflict): Conflict {
        if (child in debugChildren) {
            Lok.debug("nein!!!222e")
            return this
        }
        children.add(child)
        child.parent = this
        return this
    }

    fun decideRemote(): Conflict {
        this.decision = remoteStage
        rejection = localStage
        children.forEach { it.decideRemote() }
        return this
    }


    fun decideLocal(): Conflict {
        this.decision = localStage
        rejection = remoteStage
        children.forEach { it.decideLocal() }
        return this
    }

    override fun toString(): String = "key: \"$key\", l: \"${localStage?.name ?: "null"}\", r: \"${
        remoteStage?.name
            ?: "null"
    }\""

    fun decideNothing(): Conflict {
        this.decision = null
        rejection = null
        children.forEach { it.decideNothing() }
        return this
    }

    fun hasLocalStage(): Boolean = this.localStage != null

    fun hasRemoteStage(): Boolean = this.remoteStage != null
    fun hasParent(): Boolean = this.parent != null
    fun hasChildren(): Boolean = !this.children.isEmpty()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Conflict

        if (conflictDao != other.conflictDao) return false
        if (localStage != other.localStage) return false
        if (remoteStage != other.remoteStage) return false
        if (localId != other.localId) return false
        if (remoteId != other.remoteId) return false
        if (children != other.children) return false

        return true
    }

    override fun hashCode(): Int {
        var result = localStage?.hashCode() ?: 0
        result = 31 * result + (remoteStage?.hashCode() ?: 0)
        result = 31 * result + (localId?.hashCode() ?: 0)
        result = 31 * result + (remoteId?.hashCode() ?: 0)
        return result
    }

    companion object {
        fun createKey(lStage: Stage?, rStage: Stage?): String =
            "${lStage?.id?.toString() ?: "n"}/${rStage?.id?.toString() ?: "n"}"

        val stage = Stage()

        fun getRootConflicts(conflicts: Collection<Conflict>): List<Conflict> {
            return conflicts.filter { !it.hasParent() }
        }

    }
}